jraft-core/src/main/java/com/alipay/sofa/jraft/util/ThreadPoolsFactory.java

```diff
@@ -0,0 +1,123 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.alipay.sofa.jraft.util;
+
+import com.alipay.sofa.jraft.Closure;
+import com.alipay.sofa.jraft.Status;
+import org.apache.commons.lang.StringUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.Future;
+import java.util.concurrent.SynchronousQueue;
+import java.util.concurrent.ThreadPoolExecutor;
+
+/**
+ * ThreadPool based on Raft-Group isolation
+ *
+ * @author tynan.liu
+ * @date 2022/6/24 17:25
+ **/
+public class ThreadPoolsFactory {
+    private static final Logger                                    LOG                      = LoggerFactory
+                                                                                                .getLogger(ThreadPoolsFactory.class);
+    private static final ConcurrentMap<String, ThreadPoolExecutor> GROUP_THREAD_POOL_ROUTER = new ConcurrentHashMap<>();
+
+    private static class GlobalThreadPoolHolder {
+        private static volatile ThreadPoolExecutor DEFAULT_GLOBAL_THREAD_POOL;
+
+        private GlobalThreadPoolHolder() {
+            if (DEFAULT_GLOBAL_THREAD_POOL != null) {
+                throw new IllegalStateException(
+                    "DEFAULT_GLOBAL_THREAD_POOL has been created, please avoid using reflection to create this instance.");
+            }
+        }
+
+        private static ThreadPoolExecutor getGlobalExecutor() {
+            if (DEFAULT_GLOBAL_THREAD_POOL == null) {
+                synchronized (GlobalThreadPoolHolder.class) {
+                    if (DEFAULT_GLOBAL_THREAD_POOL == null) {
+                        DEFAULT_GLOBAL_THREAD_POOL = ThreadPoolUtil.newBuilder()
+                            .poolName("JRAFT_GROUP_DEFAULT_EXECUTOR").enableMetric(true)
+                            .coreThreads(Utils.MIN_CLOSURE_EXECUTOR_POOL_SIZE)
+                            .maximumThreads(Utils.MAX_CLOSURE_EXECUTOR_POOL_SIZE).keepAliveSeconds(60L)
+                            .workQueue(new SynchronousQueue<>())
+                            .threadFactory(new NamedThreadFactory("JRaft-Group-Default-Executor-", true)).build();
+                    }
+                }
+            }
+            return DEFAULT_GLOBAL_THREAD_POOL;
+        }
+    }
+
+    /**
+     * You can specify the ThreadPoolExecutor yourself here
+     * @param groupId Raft-Group
+     * @param executor To specify ThreadPoolExecutor
+     */
+    public static void registerThreadPool(String groupId, ThreadPoolExecutor executor) {
+        if (executor == null) {
```

Comment:
I think we should throw `IllegalArgumentException` if `executor` is null. User should not call this function when the `executor` is null.